Fix deserializing multiple self-loop edges

mmarovic opened the pull request:
Fixes #2669

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

porunov commented:
@mmarovic Thank you for the fix! Could you please resolve conflicts and add a test to cover the bug?

mmarovic commented:
@porunov I added a test to cover the specific bug. Also, I changed the PR base to v0.5 because that's the original branch I based my branch on. However, this seems to cause issues with the codacity code analysis.

Should I change the base back to master and resolve conflicts or leave it as it is now?

porunov commented:
@mmarovic Thank you! `v0.5` uses Travis CI still but it's disabled now, so I'm not sure if `v0.5` will be supported. I would recommend to change back to `master` branch as we can run CI only on `master` branch right now (using GitHub Actions).

porunov commented:
@mmarovic I think you probably need to add a copy of your test `testReadMultipleSelfEdges` to `HBaseSnapshotInputFormatIT` and overwrite it. I guess you you will need to add a line:
```
        // Take a snapshot of the graph table
        hBaseContainer.createSnapshot(snapshotName, table);
```

similarly as it is in `testReadSelfEdge` (didn't check that).

P.S. That's definitely against DRY principle and we should refactor that but no need to do such refactoring in your PR.

porunov commented on janusgraph-hadoop/src/main/java/org/janusgraph/hadoop/formats/util/JanusGraphVertexDeserializer.java:
```diff
@@ -59,14 +60,14 @@ public JanusGraphVertexDeserializer(final JanusGraphHadoopSetup setup) {
         this.idManager = setup.getIDManager();
     }
 
+    private Boolean edgeExists(Vertex vertex, RelationCache possibleDuplicate) {
```
Not need to return wrapper type. I would simply return `boolean` instead of `Boolean` here.

mmarovic replied:
I originally left it this way because the previous implemnetation also returned the wrapper type. However, I agree with you and will change it.

porunov commented on janusgraph-hadoop/src/main/java/org/janusgraph/hadoop/formats/util/JanusGraphVertexDeserializer.java:
```diff
@@ -59,14 +60,14 @@ public JanusGraphVertexDeserializer(final JanusGraphHadoopSetup setup) {
     }
 
+    private Boolean edgeExists(Vertex vertex, RelationCache possibleDuplicate) {
+        final RelationType type = typeManager.getExistingRelationType(possibleDuplicate.typeId);
```
We already computed this `type` before, so no need to compute it again. Instead you could pass this `type` into this method. I.e. could be something like `private boolean edgeExists(Vertex vertex, RelationType type, RelationCache possibleDuplicate)`

mmarovic replied:
True, I'll change this as well.

porunov replied:
Thank you! I see that the latest force push didn't update the commit, so I unresolved this commits

porunov reviewed (approved):
LGTM. Thank you @mmarovic ! The fix makes sense to me I just added two optional nitpicks

li-boxuan reviewed (approved):
LGTM 👍  Thank you for the fix!
